### ISSUE TITLE
The first value of preprocessing.text.Tokenizer texts_to_matrix() is always zero

### ISSUE THREAD

**Issue by Hoiy** [keras-team/keras#8583]
Could be reproduced by below, looks like there is an index shifted
```
from keras.preprocessing.text import Tokenizer

tokenizer = Tokenizer(num_words = 3)
texts = ['a b b c c c', 'a b c']
tokenizer.fit_on_texts(texts)
print(tokenizer.texts_to_matrix(texts, mode='count'))
```
Output:
```
[[ 0.  3.  2.]
 [ 0.  1.  1.]]
```
Expected (?):
```
[[ 3. 2. 1.]
 [ 1. 1. 1.]]
```

**Comment by nicolewhite** [comment]
Yeah the Tokenizer class seems to have been written under the assumption that you'll be using it with an [Embedding layer](https://keras.io/layers/embeddings/#embedding), where the Embedding layer expects `input_dim` to be vocab size + 1. The Tokenizer reserves the 0 index for masking (even though the default for the Embedding layer is `mask_zero=False`...), so you are only actually tokenizing the top `num_words - 1` words. This is convenient if you **are** using the Embedding layer, because then you can use the same number for `num_words` and `input_dim`. But if you're not using the Embedding layer it is just an incorrect implementation.

On my personal fork of Keras I've added a `reserve_zero` argument so that you can turn this behavior off and actually tokenize the top `num_words`:

```python
from keras.preprocessing.text import Tokenizer

tokenizer = Tokenizer(num_words=3, reserve_zero=False)
texts = ['a b b c c c', 'a b c']
tokenizer.fit_on_texts(texts)
print(tokenizer.texts_to_matrix(texts, mode='count'))
```

```
[[ 3.  2.  1.]
 [ 1.  1.  1.]]
```

A workaround for you is:

```python
from keras.preprocessing.text import Tokenizer

tokenizer = Tokenizer(num_words=4)
texts = ['a b b c c c', 'a b c']
tokenizer.fit_on_texts(texts)
matrix = tokenizer.texts_to_matrix(texts, mode='count')
matrix[:, 1:]
```

```
[[ 3.,  2.,  1.]
 [ 1.,  1.,  1.]]
```

=(

Will open a PR with the `reserve_zero` arg if enough people want it.

**Comment by asos-elaine** [comment]
Yes please, this really confused me too!
